### DISCUSSION TITLE
Support for the modular Firebase 9

### DISCUSSION THREAD

**Discussion by atanaskanchev** [nextauthjs/next-auth#3956] [Ideas]
The modular Firebase v9 is in the final beta stages and it's quite stable. I am looking forward to adding support for it to the Firebase adapter.

**Comment by balazsorban44** [comment]
I don't oppose it, but since the adapter is server-side only, it won't really have a big effect in my opinion. Or is there any other reason for version 9 to be better than modularity?

**Comment by atanaskanchev** [comment]
The main shtick of v9 is the reduced bundle size on the both client/server side. This is what they say

> The fully modularized version 9 has these advantages over earlier versions:
> 
>     Version 9 enables a dramatically reduced app size. It adopts the modern JavaScript Module format, allowing for "tree shaking" practices in which you import only the artifacts your app needs. Depending on your app, tree-shaking with version 9 can result in 80% less kilobytes than a comparable app built using version 8.
>     Version 9 will continue to benefit from ongoing feature development, while version 8 will be frozen at a point in the future.

**Comment by wobsoriano** [comment]
v9 is gonna be a refactor I suppose. Things like:

```typescript
async createUser(profile) {
  const userRef = await client.collection("users").add(newUser)
  const snapshot = await userRef.get()
  const user = docSnapshotToObject(snapshot)
  return user
}
```

is going to be

```typescript
import { collection, doc, addDoc } from "firebase/firestore"; 

async createUser(profile) {
  const userRef = await addDoc(collection("users", newUser));
  const snapshot = await getDoc(userRef)
  const user = docSnapshotToObject(snapshot)
  return user
}
```

**Comment by balazsorban44** [comment]
would like to land https://github.com/nextauthjs/next-auth/pull/2361 first, but yeah. it will need a refractor.

**Comment by balazsorban44** [comment]
@atanaskanchev, @wobsoriano the refactored version is available at nextauthjs/adapters#183. I would appreciate it if you could look and help to think through how we could utilize Firebase 9.

My biggest question is how to provide all the necessary methods because I would like to keep the adapter APIs for different adapters similar. Meaning most adapters pass a `client` object, which contains all the necessary methods. I am thinking we will have to require the users to write some boilerplate and create a `FirebaseClient` object that is passed as an argument to `FirebaseAdapter`.

Taking the above example by @wobsoriano for `createUser`, something like this:

```js
// [...nextauth].js

import { FirebaseAdapter } from "@next-auth/firebase-adapter"
import { collection, doc, addDoc } from "firebase/firestore"

const firebaseClient = { collection, doc, addDoc } // Add any necessary methods here
export default NextAuth({
  ...
  adapter: FirebaseAdapter(firebaseClient)
})
```

**Comment by wobsoriano** [comment]
Oh wait... looks like web only has the modular update. Not sure 

![Screen Shot 2021-08-17 at 7 03 29 AM](https://user-images.githubusercontent.com/13049130/129639701-e9a9b4fb-0fd8-4d9b-9e49-b9dcff4cd05d.png)

**Comment by balazsorban44** [comment]
So it might not be necessary to upgrade after all? 😁

We are using the package simply called `"firebase"`.

**Comment by wobsoriano** [comment]
Just tried `"firebase": "9.0.0-beta.7"` in a node environment and it's actually supported too.

So yeah, your example will do 😁

```typescript
import { FirebaseAdapter } from '@next-auth/firebase-adapter'
import { initializeApp } from 'firebase/app'

import {
  getFirestore,
  collection,
  query,
  getDocs,
  where,
  limit,
  doc,
  getDoc,
  addDoc,
  updateDoc,
  deleteDoc
} from 'firebase/firestore'

const firebaseApp = initializeApp({ /* config */ });

const db = getFirestore(firebaseApp)

const firebaseClient = {
  db,
  collection,
  query,
  getDocs,
  where,
  limit,
  doc,
  getDoc,
  addDoc,
  updateDoc,
  deleteDoc
}

export default NextAuth({
  adapter: FirebaseAdapter(firebaseClient)
})
```

```typescript
// somewhere
async createUser(profile) {
  const userRef = await addDoc(collection(db, 'users'), newUser);
  const snapshot = await getDoc(userRef)
  const user = docSnapshotToObject(snapshot)
  return user
}
```

**Comment by balazsorban44** [comment]
Thanks, I will have a look then!

The question is, are people ready for Firebase 9 yet, or should we publish the next version with 8, and when 9 gets to stable, we refactor then?

I don't want to maintain 2 different versions of the adapter.

**Comment by balazsorban44** [comment]
@atanaskanchev, @wobsoriano I did the rewrite, `@next-auth/firebase-adapter@0.1.2-pr.183-638762fe.8` requires at least Firebase 9. All tests pass, but could any of you test it out? You will need `next-auth@4.0.0-next.23` at least.

**Comment by abhinav-TB** [comment]
@balazsorban44 I am facing some issues while using firebase 9, what I passed into the firebase adapter function is the instance of firestore  returned by geFirestore()

the error I am getting is 
![image](https://user-images.githubusercontent.com/54744701/139122514-176d0020-4812-4ac3-ab6a-24d4f54c1e33.png)


here is my next_auth.js file
![image](https://user-images.githubusercontent.com/54744701/139122355-c7d36a18-1852-4f7a-b693-1e89e611bd9f.png)

firebase config
![image](https://user-images.githubusercontent.com/54744701/139122438-cf6df1d0-293e-480d-b7c2-5ba5745d50e2.png)

**Comment by ceevie** [comment]
> @atanaskanchev, @wobsoriano I did the rewrite, `@next-auth/firebase-adapter@0.1.2-pr.183-638762fe.8` requires at least Firebase 9. All tests pass, but could any of you test it out? You will need `next-auth@4.0.0-next.23` at least.

Will there be a new official update to the Firebase Adapter Docs?
Because the current version of the docs isn't compatible with firebase v9.
I had to add the FirebaseClient like mentioned above in order to work.

**Comment by balazsorban44** [comment]
Still haven't decided. See nextauthjs/adapters#183 What's the status of Firebase 9? Is it stable yet?

**Comment by michalscislowski** [comment]
```import { FirebaseAdapter } from '@next-auth/firebase-adapter'
import { initializeApp } from 'firebase/app'

import {
  getFirestore,
  collection,
  query,
  getDocs,
  where,
  limit,
  doc,
  getDoc,
  addDoc,
  updateDoc,
  deleteDoc
} from 'firebase/firestore'

const firebaseApp = initializeApp({ /* config */ });

const db = getFirestore(firebaseApp)

const firebaseClient = {
  db,
  collection,
  query,
  getDocs,
  where,
  limit,
  doc,
  getDoc,
  addDoc,
  updateDoc,
  deleteDoc
}

export default NextAuth({
  adapter: FirebaseAdapter(firebaseClient)
})
```

> ```ts
> async createUser(profile) {
>   const userRef = await addDoc(collection(db, 'users'), newUser);
>   const snapshot = await getDoc(userRef)
>   const user = docSnapshotToObject(snapshot)
>   return user
> }
> ```

I'm getting ```FirebaseError: Expected first argument to collection() to be a CollectionReference, a DocumentReference or FirebaseFirestore``` this way. Anyone had similar issue?

**Comment by greendesertsnow** [comment]
Yes Firebase is stable. It's version ^9.4.1

> Still haven't decided. See nextauthjs/adapters#183 What's the status of Firebase 9? Is it stable yet?

**Comment by balazsorban44** [comment]
We should go for v9 then I think. I'm not able to work on the PR, so if anyone is interested in picking up, please do 🙏. Would be awesome to figure out if we can integrate with Firebase Auth as well.

**Comment by SahilAujla** [comment]
Please add support for v9 as soon as possible. It's a humble request.

**Comment by balazsorban44** [comment]
"as soon as possible" is not really humble. :confused: Please read nextauthjs/next-auth#3827. Feel free to work on it. 

Useful links:

https://next-auth.js.org/adapters/overview
https://next-auth.js.org/adapters/models
https://next-auth.js.org/tutorials/creating-a-database-adapter

See other adapters for example on testing, TypeScript etc.:
https://github.com/nextauthjs/adapters/tree/main/packages

**Comment by SahilAujla** [comment]
I am not good with words. I am sorry if that was rude in any way. It's just I am stuck on a project where I am using Firebase v9. 

It's my first time using any adapter, I don't even know what an adapter is properly, I was just following a tutorial. 

But I will definitely try to work on it, whatever I can.

**Comment by balazsorban44** [comment]
No hard feelings! :green_heart: 

Hopefully, the above links should give you a proper introduction. When you've read through all of them and still have questions, don't hesitate to ask.

**Comment by srijans38** [comment]
Please see https://github.com/nextauthjs/adapters/issues/180#issuecomment-1004126562 with regards to firebase auth. The issue seems to be closed. I have also got it in a basic working state here nextauthjs/adapters#363

**Comment by greendesertsnow** [comment]
React-Query-Firebase is pretty much working like a firebase adapter right now @srijans38 @SahilAujla
I'm using it until I see a stable fb v9 for next-auth.
I'm sure there's things next-auth does better with SSR by the way.

**Comment by ndom91** [comment]
`nextauthjs/next-auth` has been migrated to a monorepository. The adapters code can now be found there under `packages/adapter-*`. Thanks for your interest in the project!
